# Biostat M280 Homework 1

#### Due Apr 27 @ 11:59PM

## Q1

**No handwritten homework reports are accepted for this course.**  We work with Git/GitHub.  Efficient and abundant use of Git, e.g., **frequent and well-documented** commits, is an important criterion for grading your homework.

0. Apply for the [Student Developer Pack](https://education.github.com/pack) at GitHub using your UCLA email.

0. Create a **private** repository `biostat-m280-2018-spring` and add `Hua-Zhou` and `LuZhangstat` (TA) as your collaborators.

0. Top directories of the repository should be `hw1`, `hw2`, ... Create two branches `master` and `develop`. The `develop` branch will be your main playground, the place where you develop solution (code) to homework problems and write up report. The `master` branch will be your presentation area. Put your homework submission files (IJulia notebook `.ipynb`, html converted from notebook, all code and data set to reproduce results) in `master` branch.  

0. After each homework due date, teaching assistant and instructor will check out your `master` branch for grading. Tag each of your homework submissions with tag names `hw1`, `hw2`, ...  Tagging time will be used as your submission time. That means if you tag your hw1 submission after deadline, penalty points will be deducted for late submission.  

0. Read the [style guide](https://github.com/johnmyleswhite/Style.jl) for Julia programming by John Myles White. Following rules in the style guide will be strictly enforced when grading: (4), (6), (7), (8), (9), (12), (13) and (16).

***  Have done the github setup. Username: huiyuhu ***

## Q2

Let's check whether floating-point numbers obey certain algebraic rules.  
0. Associative rule for addition says `(x + y) + z == x + (y + z)`. Check association rule using `x = 0.1`, `y = 0.1` and `z = 1.0` in Julia. Explain what you find.

0. Do floating-point numbers obey the associative rule for multiplication: `(x * y) * z == x * (y * z)`?

0. Do floating-point numbers obey the distributive rule: `a * (x + y) == a * x + a * y`?  

0. Is `0 * x == 0` true for all floating-point number `x`?  

0. Is `x / a == x * (1 / a)` always true?

In [49]:
# 1
x = 0.1;
y = 0.1;
z = 1.0;

(x + y) + z == x + (y + z)

false

In [50]:
(x + y) + z

1.2

In [51]:
x + (y + z)

1.2000000000000002

* The result showed that the associative rule for addition is not ture. The reason is that x + (y + z) = 1.2000000000000002. Floating point numbers have a fixed length which means that round off errors are common. The bits showed as follow:

In [52]:
@show bits(x)  
@show bits(y)
@show bits(z)
@show bits(x + y)
@show bits(y + z)
@show bits(x + (y + z))
@show bits((x + y) + z) #bits are different

bits(x) = "0011111110111001100110011001100110011001100110011001100110011010"
bits(y) = "0011111110111001100110011001100110011001100110011001100110011010"
bits(z) = "0011111111110000000000000000000000000000000000000000000000000000"
bits(x + y) = "0011111111001001100110011001100110011001100110011001100110011010"
bits(y + z) = "0011111111110001100110011001100110011001100110011001100110011010"
bits(x + (y + z)) = "0011111111110011001100110011001100110011001100110011001100110100"
bits((x + y) + z) = "0011111111110011001100110011001100110011001100110011001100110011"


"0011111111110011001100110011001100110011001100110011001100110011"

In [61]:
# 2
x = 0.1;
y = 0.2;
z = 1.3;
(x * y) * z == x * (y * z)

false

* Let x = 0.1, y =0.2, z = 0.3, the associative rule for multiplication is false. The bitscode showed as follow: 

In [62]:
@show bits(x)  
@show bits(y)
@show bits(z)
@show bits(x * y)
@show bits(y * z)
@show bits(x * (y * z))
@show bits((x * y) * z) #bits are different

bits(x) = "0011111110111001100110011001100110011001100110011001100110011010"
bits(y) = "0011111111001001100110011001100110011001100110011001100110011010"
bits(z) = "0011111111110100110011001100110011001100110011001100110011001101"
bits(x * y) = "0011111110010100011110101110000101000111101011100001010001111100"
bits(y * z) = "0011111111010000101000111101011100001010001111010111000010100100"
bits(x * (y * z)) = "0011111110011010100111111011111001110110110010001011010000111010"
bits((x * y) * z) = "0011111110011010100111111011111001110110110010001011010000111011"


"0011111110011010100111111011111001110110110010001011010000111011"

In [69]:
# 3
a = 0.000001;
a * (x + y) == a * x + a * y

false

* Let a = 0.000001, distributive rule is not true.

In [70]:
@show bits(a * (x + y))
@show bits(a * x + a * y) # bits are different

bits(a * (x + y)) = "0011111010010100001000011111010111110100000011011000001101110111"
bits(a * x + a * y) = "0011111010010100001000011111010111110100000011011000001101110110"


"0011111010010100001000011111010111110100000011011000001101110110"

In [53]:
# 4
0 * x == 0

true

In [55]:
0 * Inf == 0
@show 0 * Inf

0Inf = NaN


NaN

* If x is a infinite number, 0 * x is not true.

In [73]:
# 5 
x = 3.0; a = 10.0;
x / a == x * (1 / a) # bits are different

false

* Let x = 3.0, a = 10.0, x / a = x * (1 / a) is not true.

In [72]:
@show bits(x / a)
@show bits(x * (1 / a)) # bits are different

bits(x / a) = "0011111111010011001100110011001100110011001100110011001100110011"
bits(x * (1 / a)) = "0011111111010011001100110011001100110011001100110011001100110100"


"0011111111010011001100110011001100110011001100110011001100110100"

## Q3

Consider Julia function
```julia
function g(k)
    for i = 1:10
        k = 5k - 1
    end
    k
end
```
0. Use `@code_llvm` to find the LLVM bitcode of compiled `g` with `Int64` input.   
1. Use `@code_llvm` to find the LLVM bitcode of compiled `g` with `Float64` input.  
2. Compare the bitcode from questions 1 and 2. What do you find?  
3. Read Julia documentation on `@fastmath` and repeat the questions 1-3 on the function  

```julia
function g_fastmath(k)  
    @fastmath for i = 1:10  
        k = 5k - 1
    end
    k
end
```
Explain what does macro `@fastmath` do?

In [74]:
function g(k)
    for i = 1:10
        k = 5k - 1
    end
    k
end

g (generic function with 1 method)

Use 1 as an Int64 input and 1.0 as a Float64 input.

In [77]:
@code_llvm g(1)


define i64 @julia_g_63201(i64) #0 !dbg !5 {
top:
  %1 = mul i64 %0, 9765625
  %2 = add i64 %1, -2441406
  ret i64 %2
}


In [78]:
@code_llvm g(1.0)


define double @julia_g_63203(double) #0 !dbg !5 {
top:
  %1 = fmul double %0, 5.000000e+00
  %2 = fadd double %1, -1.000000e+00
  %3 = fmul double %2, 5.000000e+00
  %4 = fadd double %3, -1.000000e+00
  %5 = fmul double %4, 5.000000e+00
  %6 = fadd double %5, -1.000000e+00
  %7 = fmul double %6, 5.000000e+00
  %8 = fadd double %7, -1.000000e+00
  %9 = fmul double %8, 5.000000e+00
  %10 = fadd double %9, -1.000000e+00
  %11 = fmul double %10, 5.000000e+00
  %12 = fadd double %11, -1.000000e+00
  %13 = fmul double %12, 5.000000e+00
  %14 = fadd double %13, -1.000000e+00
  %15 = fmul double %14, 5.000000e+00
  %16 = fadd double %15, -1.000000e+00
  %17 = fmul double %16, 5.000000e+00
  %18 = fadd double %17, -1.000000e+00
  %19 = fmul double %18, 5.000000e+00
  %20 = fadd double %19, -1.000000e+00
  ret double %20
}


In [79]:
function g_fastmath(k)  
    @fastmath for i = 1:10  
        k = 5k - 1
    end
    k
end

g_fastmath (generic function with 1 method)

In [82]:
@code_llvm g_fastmath(1)


define i64 @julia_g_fastmath_63226(i64) #0 !dbg !5 {
top:
  %1 = mul i64 %0, 9765625
  %2 = add i64 %1, -2441406
  ret i64 %2
}


In [83]:
@code_llvm g_fastmath(1.0)


define double @julia_g_fastmath_63227(double) #0 !dbg !5 {
top:
  %1 = fmul fast double %0, 9.765625e+06
  %2 = fadd fast double %1, -2.441406e+06
  ret double %2
}


## Q4

Create the vector `x = (0.988, 0.989, 0.990, ..., 1.010, 1.011, 1.012)`.  
0. Plot the polynomial `y = x^7 - 7x^6 + 21x^5 - 35x^4 + 35x^3 - 21x^2 + 7x -1` at points `x`.  

1. Plot the polynomial `y = (x - 1)^7` at points `x`.  

2. Explain what you found.

In [93]:
#Install package
#Pkg.add("Plots")
#Pkg.add("PyPlot")

In [94]:
# Creat the vector x
x = 0.988:0.002:1.012; 
y1 = x.^7 - 7x.^6 + 21x.^5 - 35x.^4 + 35x.^3 - 21x.^2 + 7x - 1;
using Plots
pyplot()

plot(x, y1)

if !Plots.is_installed("PyPlot")
    Pkg.add("PyPlot")
end
withenv("PYTHON" => "") do
    Pkg.build("PyPlot")
end

# now restart julia!




INFO: To do a standard install of pyplot, copy and run this:



LoadError: ArgumentError: Module PyPlot not found in current path.
Run `Pkg.add("PyPlot")` to install the PyPlot package.

## Q5

Let the $n \times n$ matrix `H` have elements `H[i, j] = 1 / (i + j - 1)`.  
0. Write a function `h(n)` that outputs $n \times n$ matrix `H`. Try at least 3 ways, e.g., looping, [comprehension](https://docs.julialang.org/en/stable/manual/arrays/#Comprehensions-1), and vectorization. Compute and print `H` for `n = 5`.   
0. Compare their efficiencies at `n = 1000`.
0. Try different rounding modes (use function `setrounding(Float64, RoundingMode)`) and report the entry `inv(H)[1, 1]` for `n = 15`.

## Q6

0. Show the **Sherman-Morrison formula**
$$
	(\mathbf{A} + \mathbf{u} \mathbf{u}^T)^{-1} = \mathbf{A}^{-1} - \frac{1}{1 + \mathbf{u}^T \mathbf{A}^{-1} \mathbf{u}} \mathbf{A}^{-1} \mathbf{u} \mathbf{u}^T \mathbf{A}^{-1},
$$
where $\mathbf{A} \in \mathbb{R}^{n \times n}$ is nonsingular and $\mathbf{u} \in \mathbb{R}^n$. This formula supplies the inverse of the symmetric, rank-one  perturbation of $\mathbf{A}$.

0. Show the **Woodbury formula**
$$
	(\mathbf{A} + \mathbf{U} \mathbf{V}^T)^{-1} = \mathbf{A}^{-1} - \mathbf{A}^{-1} \mathbf{U} (\mathbf{I}_m + \mathbf{V}^T \mathbf{A}^{-1} \mathbf{U})^{-1} \mathbf{V}^T \mathbf{A}^{-1},
$$
where $\mathbf{A} \in \mathbb{R}^{n \times n}$ is nonsingular, $\mathbf{U}, \mathbf{V} \in \mathbb{R}^{n \times m}$, and $\mathbf{I}_m$ is the $m \times m$ identity matrix. In many applications $m$ is much smaller than $n$. Woodbury formula generalizes Sherman-Morrison and is valuable because the smaller matrix $\mathbf{I}_m + \mathbf{V}^T \mathbf{A}^{-1} \mathbf{U}$ is cheaper to invert than the larger matrix $\mathbf{A} + \mathbf{U} \mathbf{V}^T$.

0. Show the **binomial inversion formula**
$$
	(\mathbf{A} + \mathbf{U} \mathbf{B} \mathbf{V}^T)^{-1} = \mathbf{A}^{-1} - \mathbf{A}^{-1} \mathbf{U} (\mathbf{B}^{-1} + \mathbf{V}^T \mathbf{A}^{-1} \mathbf{U})^{-1} \mathbf{V}^T \mathbf{A}^{-1},
$$
where $\mathbf{A}$ and $\mathbf{B}$ are nonsingular.

0. Show the identity
$$
	\text{det}(\mathbf{A} + \mathbf{U} \mathbf{V}^T) = \text{det}(\mathbf{A}) \text{det}(\mathbf{I}_m + \mathbf{V}^T \mathbf{A}^{-1} \mathbf{U}).
$$
This formula is useful for evaluating the density of a multivariate normal with covariance matrix $\mathbf{A} + \mathbf{U} \mathbf{V}^T$.